### PR TITLE
Feat: callbacks

### PR DESCRIPTION
**Pull request by liyier90**
Enables users to attach callback functions to nodes to be run at certain points/events in the pipeline.

**Implemented events**:
- `run_begin`: Before `Node.run()`
- `run_end`: After `Node.run()`

**Supported callback types**:
- Pure function
- Class methods
- Static methods
- Instance methods (require instantiating the object in the callback script)

**UI**:
Users will specify callbacks for the node under the `callbacks` config key. Each callback is specified as:
```
callback definition = module part, "::", callback part
module part         = { subdir name, "." }, script name
callback part       = function
                     | class name, "::", method
                     | instance name, "::", instance method

script name     = ? name of callback Python script ?
subdir name     = ? name of individual subdirectory folders relative to `callbacks` folder ?
class name      = ? name of class which contains the callback methods ?
instance name   = ? name of instantiated object which contains the callback method ?
function        = ? name of callback function ?
method          = ? name of class/static methods ?
instance method = ? name of instance method ?
```

Incorrect callback definitions are skipped with a warning log, program is not halted.

**Implementation detail**:
In `_edit_config()` of both `AbstractNode` and `DeclarativeLoader`, avoid going deeper into nested dictionary if the key is `callbacks`. This allows us to define the callbacks config in PeekingDuck as
```
callbacks: {}
```
and in pipeline config as
```
- node.name:
    callbacks:
      run_begin: ["callback::definition"]
```
This allows us to implement more callback events in the future without have to add the event key to all the config files in PeekingDuck.

**Sample usage**:
Extract [callbacks.zip](https://github.com/aisingapore/PeekingDuck/files/9771643/callbacks.zip) into your project directory. Ensure that the `callbacks` directory is at the same level as `dev_callback.yml`. You should have the following directory layout:
```
project_dir
+-callbacks
| +-cb_dir
| +-my_callbacks.py
| \-sql_callback.py
+-dev_callback.yml
\-...
```

Run `dev_callback.yml` and you should see logging outputs such as
- `called 1` from `callback_1()` in `my_callbacks.py`
- `called z` from `cb_1()` in `cb_dir/mycb.py`
- `Suppressing logging` from `MyCallback.callback_2()` in `my_callbacks.py`
- `Exceed zone limit ...` from `MyCallback.callback_3()` in `my_callbacks.py`
- `Update db with frame_count ...` from `cb_obj.track_count_callback()` in `sql_callback.py`

**Additional changes**:
- Refactored depracation code into separate methods (for ease of removal in the future)
- Removed `key` argument from `obj_det_class_name_to_id()`

### CODE REVIEW

**Comment by ongtw** [comment]
Can you write up a tutorial `Using Callbacks` under the `Peaking Duck` tutorial section on using callbacks and giving some sample codes/pipelines as illustrations?

**Comment by liyier90** [comment]
Added a "Using Callbacks" section to Peaking Duck tutorial. Includes:
- tutorial to go over the implemented events and callback definition
- tutorial to replicate the "Interfacing with SQL" tutorial, replacing the `output.sqlite` custom node with a callback

File: peekingduck/nodes/callback_list.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+    @classmethod
+    def from_dict(cls, callback_dict: Dict[str, List[str]]) -> "CallbackList":
+        """Constructs a `CallbackList` object populated with callbacks from
+        `callback_dict`.
+
+        Args:
+            callback_dict (Dict[str, List[str]]): A dictionary defined in the
+                pipeline config file. Maps `event_type` to a list of callback
+                definitions. Each callback defintion should contain:
```
Comment (ongtw) [inline peekingduck/nodes/callback_list.py]: Typo: `defintion`

File: peekingduck/nodes/callback_list.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    """
+
+    def _getattr(obj: object, name: str) -> Any:
+        return getattr(obj, name, *args)
+
+    return functools.reduce(_getattr, names, obj)
+
+
+def import_module_from_file_location(module_name: str, location: Path) -> ModuleType:
+    """Imports module from the specified location without haivng to modify
```
Comment (ongtw) [inline peekingduck/nodes/callback_list.py]: Typo: `haivng`

File: docs/source/tutorials/04_peaking_duck.rst [hunk 3]
```diff
@@ -580,3 +580,576 @@ Do a :greenbox:`peekingduck run` and you will see the following display:
...
+      subdir name     = ? name of individual subdirectory folders relative to "callbacks" folder ?
+      class name      = ? name of class which contains the callback methods ?
+      instance name   = ? name of instantiated object which contains the callback method ?
+      function        = ? name of callback function ?
+      method          = ? name of class/static methods ?
+      instance method = ? name of instance method ? 
+
+The ``callbacks`` directory, which houses the scripts containing the callback functions,
+is expected to be at the same location of the pipeline config file. For example,
+given the following directory structure and file contents:
```
Comment (ongtw) [inline docs/source/tutorials/04_peaking_duck.rst:656]: We should specify that for `instance_name::instance_method`, the instance object has to be in the script's global scope (c.f. line 692 below).
For example, if `class A` has a method `A.f()` and an instance object `obj` is created within `A.f()`, then `obj::obj_method` would not work as `obj` cannot be "seen".

File: peekingduck/utils/detect_id_mapper.py [hunk 4]
```diff
@@ -111,4 +108,4 @@ def obj_det_change_class_name_to_id(
         x if isinstance(x, int) else class_id_map.get(x, 0) for x in value_lc
     }
     obj_ids_sorted_list = sorted(list(obj_ids_set))
-    return key, obj_ids_sorted_list
+    return obj_ids_sorted_list
```
Comment (ongtw) [inline peekingduck/utils/detect_id_mapper.py:111]: Need to update function return signature in line 84 above.

File: docs/source/tutorials/04_peaking_duck.rst [hunk 5]
```diff
@@ -580,3 +580,576 @@ Do a :greenbox:`peekingduck run` and you will see the following display:
...
+   #. ``run_begin``: Invoked before a ``Node``'s ``run()`` method.
+   #. ``run_end``: Invoked after a ``Nodes``'s ``run()`` method.
+
+For example, in the following pipeline config:
+
+   .. code-block:: yaml
+      :linenos:
+
+      nodes:
+      - input.visual:
```
Comment (ongtw) [inline docs/source/tutorials/04_peaking_duck.rst]: Remove ending comma

File: docs/source/tutorials/04_peaking_duck.rst [hunk 6]
```diff
@@ -580,3 +580,576 @@ Do a :greenbox:`peekingduck run` and you will see the following display:
...
+      - model.posenet
```
Comment (ongtw) [inline docs/source/tutorials/04_peaking_duck.rst]: Need ending comma.

File: docs/source/tutorials/04_peaking_duck.rst [hunk 7]
```diff
@@ -580,3 +580,576 @@ Do a :greenbox:`peekingduck run` and you will see the following display:
...
+          callbacks:
+            run_begin: [<callback 1>]
+            run_end: [<callback 2>]
```
Comment (ongtw) [inline docs/source/tutorials/04_peaking_duck.rst]: Can change above to `[<callback 2> <callback 3> <callback 4>]` to explicitly illustrate that we support a list of callbacks.

File: docs/source/tutorials/04_peaking_duck.rst [hunk 8]
```diff
@@ -580,3 +580,576 @@ Do a :greenbox:`peekingduck run` and you will see the following display:
...
+              [
+                "my_callback::my_callback_function",
+                "my_callback::MyCallbackClass::my_callback_static_method",
+                "my_callback::MyCallbackClass::my_callback_class_method",
+                "my_callback::my_callback_obj::my_callback_static_method",
+                "my_callback::my_callback_obj::my_callback_class_method",
+                "my_callback::my_callback_obj::my_callback_instance_method",
+                "cb_dir.my_cb::my_cb_function",
+              ]
+      - output.screen
```
Comment (ongtw) [inline docs/source/tutorials/04_peaking_duck.rst:743]: The word `callback` appears many times in different forms, which can be hard to read and confusing.
Would like to suggest refactoring `my_callback_function`, `my_callback_static_method`, `my_callback_class_method`, `my_callback_instance_method` as shown in below image.
Also added some code to access and print `data_pool` contents.

image

Will have to update the following table accordingly.

**Review by ongtw — CHANGES_REQUESTED** [review]
Good work on pipeline callback.
Have some suggestions on improving the tutorial.

File: peekingduck/nodes/callback_list.py (hunk 1 above)
Comment (liyier90) [reply peekingduck/nodes/callback_list.py]: Fixed

File: peekingduck/nodes/callback_list.py (hunk 2 above)
Comment (liyier90) [reply peekingduck/nodes/callback_list.py]: Fixed

File: peekingduck/utils/detect_id_mapper.py (hunk 4 above)
Comment (liyier90) [reply peekingduck/utils/detect_id_mapper.py:111]: Fixed

File: docs/source/tutorials/04_peaking_duck.rst (hunk 3 above)
Comment (liyier90) [reply docs/source/tutorials/04_peaking_duck.rst:656]: Added clarification, also added comment in sample code.

File: docs/source/tutorials/04_peaking_duck.rst (hunk 5 above)
Comment (liyier90) [reply docs/source/tutorials/04_peaking_duck.rst]: Removed

File: docs/source/tutorials/04_peaking_duck.rst (hunk 6 above)
Comment (liyier90) [reply docs/source/tutorials/04_peaking_duck.rst]: Added

File: docs/source/tutorials/04_peaking_duck.rst (hunk 7 above)
Comment (liyier90) [reply docs/source/tutorials/04_peaking_duck.rst]: Changed

File: docs/source/tutorials/04_peaking_duck.rst (hunk 8 above)
Comment (liyier90) [reply docs/source/tutorials/04_peaking_duck.rst:743]: Updated

File: peekingduck/nodes/callback_list.py (hunk 1 above)
Comment (ongtw) [reply peekingduck/nodes/callback_list.py]: Looks like the changes have not been saved. Once this is corrected, can merge.

Comment (liyier90) [reply peekingduck/nodes/callback_list.py]: Oh sorry, I misunderstood this, changed `definitions -> definition` initally.

Corrected `defintion -> definition`

**Pull request merged**